File: examples/examples/pid_jerk_limited/run_example.cpp
// Copyright 2025 Universidad Politécnica de Madrid
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file run_example.cpp
 * @brief Unified example: PidGeometricController + JerkLimitedGenerator.
 *
 * @author Rafael Perez-Segui <r.psegui@upm.es>
 */

#include <cstdlib>
#include <iostream>
#include <memory>
#include <string>

#include "adapters/controllers/pid_geometric_controller.hpp"
#include "adapters/trajectory_generators/jerk_limited_generator.hpp"
#include "framework/waypoints_simulator.hpp"
#include "mav_simulator/simulator_yaml.hpp"
#include "utils/example_config_utils.hpp"

namespace {

struct Args {
  std::string example_config_path    = "configs/simulation/config_example.yaml";
  std::string simulator_config_path  = "configs/simulation/config_simulator.yaml";
  std::string controller_config_path = "configs/controllers/config_pid.yaml";
  std::string trajectory_config_path = "configs/generators/config_jerk_limited.yaml";
  std::string output_file            = "simulator_logs/pid_jerk_limited_log.csv";
};

Args parseArgs(int argc, char** argv) {
  Args args;
  for (int i = 1; i < argc; ++i) {
    const std::string a = argv[i];
    if ((a == "-c" || a == "--example_config") && i + 1 < argc) {
      args.example_config_path = argv[++i];
    } else if ((a == "-s" || a == "--simulator_config") && i + 1 < argc) {
      args.simulator_config_path = argv[++i];
    } else if ((a == "-k" || a == "--controller_config") && i + 1 < argc) {
      args.controller_config_path = argv[++i];
    } else if ((a == "-t" || a == "--trajectory_config") && i + 1 < argc) {
      args.trajectory_config_path = argv[++i];
    } else if ((a == "-f" || a == "--output_file") && i + 1 < argc) {
      args.output_file = argv[++i];
    } else if (a == "-h" || a == "--help") {
      std::cout << "Usage: " << argv[0] << "\n"
                << "  -c, --example_config     <yaml> (default: " << args.example_config_path << ")\n"
                << "  -s, --simulator_config   <yaml> (default: " << args.simulator_config_path << ")\n"
                << "  -k, --controller_config  <yaml> (default: " << args.controller_config_path << ")\n"
                << "  -t, --trajectory_config  <yaml> (default: " << args.trajectory_config_path << ")\n"
                << "  -f, --output_file        <csv>  (default: " << args.output_file << ")\n";
      std::exit(0);
    }
  }
  args.output_file = mpc_examples::detail::normalizeOutputPath(args.output_file);
  return args;
}

}  // namespace

int main(int argc, char** argv) {
  using namespace mpc_examples;

  const Args args                 = parseArgs(argc, argv);
  const ExampleConfig example_cfg = loadExampleConfig(args.example_config_path);
  const auto sim_params =
      mav_simulator::loadSimulatorParametersFromYaml(args.simulator_config_path);

  const auto controller_cfg =
      adapters::PidGeometricController::loadConfigFromYaml(args.controller_config_path);
  const auto traj_cfg =
      adapters::JerkLimitedGenerator::loadConfigFromYaml(args.trajectory_config_path);

  auto controller = std::make_unique<adapters::PidGeometricController>(controller_cfg);
  auto traj_gen   = std::make_unique<adapters::JerkLimitedGenerator>(traj_cfg);

  framework::WaypointsSimulator simulator(std::move(controller), std::move(traj_gen),
                                          example_cfg, sim_params, args.output_file);
  simulator.run();
  simulator.printBenchmark();
  return 0;
}

File: examples/examples/pid_dynamic/run_example.py
#!/usr/bin/env python3

# Copyright 2025 Universidad Politécnica de Madrid
# SPDX-License-Identifier: BSD-3-Clause

"""Unified Python example: PidGeometricController + DynamicTrajectoryGenerator.

Python mirror of ``examples/examples/pid_dynamic/run_example.cpp``. Consumes the
same YAML configuration files as the C++ binary so that a back-to-back CSV
comparison is possible.
"""

__authors__ = 'Rafael Perez-Segui'
__license__ = 'BSD-3-Clause'

import os
import sys

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_EXAMPLES_ROOT = os.path.dirname(_THIS_DIR)
if _EXAMPLES_ROOT not in sys.path:
    sys.path.insert(0, _EXAMPLES_ROOT)

from _common import parse_full_args  # noqa: E402

from mavpy.simulator import load_simulator_parameters_from_yaml
from mpc_examples_adapters.controllers import PidGeometricController
from mpc_examples_adapters.trajectory_generators import DynamicTrajectoryGenerator
from mpc_examples_framework import (
    WaypointsSimulator,
    load_example_config,
    normalize_output_path,
)


def main() -> int:
    args = parse_full_args(
        prog='pid_dynamic',
        default_controller_cfg='configs/controllers/config_pid.yaml',
        default_trajectory_cfg='configs/generators/config_dynamic.yaml',
        default_output_file='simulator_logs/pid_dynamic_log.csv',
    )
    output_file = normalize_output_path(args.output_file)

    example_cfg = load_example_config(args.example_config_path)
    sim_params = load_simulator_parameters_from_yaml(args.simulator_config_path)

    controller_cfg = PidGeometricController.load_config_from_yaml(args.controller_config_path)
    traj_cfg = DynamicTrajectoryGenerator.load_config_from_yaml(args.trajectory_config_path)

    controller = PidGeometricController(controller_cfg)
    traj_gen = DynamicTrajectoryGenerator(traj_cfg)

    simulator = WaypointsSimulator(controller, traj_gen, example_cfg,
                                   sim_params, output_file)
    simulator.run()
    simulator.print_benchmark()
    return 0


if __name__ == '__main__':
    sys.exit(main())
